### PR TITLE
Optimize `groupby::scan`

### PR DESCRIPTION
**Pull request by PointKernel**
Closes https://github.com/rapidsai/cudf/issues/8522

This PR gets rid of redundant rearranging processes in `groupby::scan` if input values are presorted. Instead of a short circuit in `sort_helper`, it adds an early exit in the scan functor to avoid materializing `sorted_values`/`grouped_values` thus reducing memory footprint. This optimization brings a 1.6x speedup for presorted scan operations.

- Baseline
```
-----------------------------------------------------------------------------------------
Benchmark                                               Time             CPU   Iterations
-----------------------------------------------------------------------------------------
Groupby/BasicSumScan/1000000/manual_time            0.455 ms        0.472 ms         1388
Groupby/BasicSumScan/10000000/manual_time            8.80 ms         8.81 ms           61
Groupby/BasicSumScan/100000000/manual_time            543 ms          543 ms            1
Groupby/PreSortedSumScan/1000000/manual_time        0.217 ms        0.236 ms         3319
Groupby/PreSortedSumScan/10000000/manual_time        1.45 ms         1.47 ms          479
Groupby/PreSortedSumScan/100000000/manual_time       14.0 ms         14.0 ms           47
```
- After optimization
```
-----------------------------------------------------------------------------------------
Benchmark                                               Time             CPU   Iterations
-----------------------------------------------------------------------------------------
Groupby/BasicSumScan/1000000/manual_time            0.455 ms        0.472 ms         1393
Groupby/BasicSumScan/10000000/manual_time            8.81 ms         8.82 ms           60
Groupby/BasicSumScan/100000000/manual_time            546 ms          546 ms            1
Groupby/PreSortedSumScan/1000000/manual_time        0.129 ms        0.148 ms         5389
Groupby/PreSortedSumScan/10000000/manual_time       0.901 ms        0.921 ms          769
Groupby/PreSortedSumScan/100000000/manual_time       8.68 ms         8.70 ms           74
```

### CODE REVIEW

File: cpp/include/cudf/groupby.hpp [hunk 1]
```diff
@@ -73,6 +73,7 @@ struct aggregation_request {
 struct scan_request {
   column_view values;  ///< The elements to aggregate
   std::vector<std::unique_ptr<groupby_scan_aggregation>> aggregations;  ///< Desired aggregations
+  sorted values_pre_sorted{sorted::NO};  ///< Whether or not the values are sorted
```
Comment (jrhemstad) [inline cpp/include/cudf/groupby.hpp:76]: Why is this needed?

Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp:76]: I don't think this is correct. This could conflict with the `keys_are_sorted` information provided at construction of the `groupby` object. The keys of the groupby object and the values provided to an aggregation request must always be in the same order.

The optimization discussed in https://github.com/rapidsai/cudf/issues/8522 should only use information from `keys_are_sorted` passed at construction of `groupby`.

**Comment by revans2** [comment]
This would be nice to have work, as we use this in a number of areas.

File: cpp/src/groupby/sort/functors.hpp [hunk 2]
```diff
@@ -40,12 +40,23 @@ struct store_result_functor {
                        sort::sort_groupby_helper& helper,
                        cudf::detail::result_cache& cache,
                        rmm::cuda_stream_view stream,
-                       rmm::mr::device_memory_resource* mr)
-    : helper(helper), cache(cache), values(values), stream(stream), mr(mr)
+                       rmm::mr::device_memory_resource* mr,
+                       sorted keys_are_sorted = sorted::NO)
```
Comment (vuule) [inline cpp/src/groupby/sort/functors.hpp:44]: Would it be possible/preferable to set `sorted_values`/`grouped_values` as views to `values` in the constructor, when keys are sorted? To me it looks like this would replace multiple early exits (the fewer branches the better).
I think this would eliminate the early exit from `scan_result_functor::get_grouped_values` as well. Also eliminates the need for `is_presorted` AFAICT.

**Review by vuule — COMMENTED** [review]
Looks good, just have an idea to potentially make this even simpler.

File: cpp/src/groupby/sort/functors.hpp (hunk 2 above)
Comment (PointKernel) [reply cpp/src/groupby/sort/functors.hpp:44]: Right, we can set `sorted_values`/`grouped_values` as views of existing data like `values` if input keys are presorted. When the input data is not sorted, however, creating and materializing `sorted_values`/`grouped_values` is necessary thus I'm not sure how the non-owning view will not work in this case. Setting up `sorted_values`/`grouped_values` in the constructor when presorted can avoid early exit branches but will cause redundant materialization.

Comment (PointKernel) [reply cpp/src/groupby/sort/functors.hpp:44]: @karthikeyann any comments on your side are appreciated since you are the original author of this code.

Comment (vuule) [reply cpp/src/groupby/sort/functors.hpp:44]: Ah, I mixed it up, as if `values` is a column and `sorted_values` is a view :\
There's still a way to do what I suggested, but AFAICT it's not better than the current code.

**Comment by vuule** [comment]
Is there a benchmark that shows the perf benefit?

**Comment by devavret** [comment]
I think the original plan was better to short circuit in the sort_helper. It is also used in rolling. Otherwise, we'd have to fix it again for rolling.

Would it be possible or would it require a bigger refactor?

**Comment by PointKernel** [comment]
> I think the original plan was better to short circuit in the sort_helper. It is also used in rolling. Otherwise, we'd have to fix it again for rolling.
> 
> Would it be possible or would it require a bigger refactor?

@devavret It's definitely possible and actually involves fewer code changes than the current refactor. The question is really whether we want to materialize `sorted_values`/`grouped_values` or not. short circuit in `sort_helper` will materialize/memorize those two while the current refactor won't.

**Comment by devavret** [comment]
> short circuit in `sort_helper` will materialize/memorize those two while the current refactor won't (thus less memory footprint).

You can memoize both the view and the owning column. Anyway I now think your method would be cleaner

**Comment by PointKernel** [comment]
@vuule I've just added the benchmark results. This PR brings about 60% speedups for presorted scans.

**Review by vuule — APPROVED** [review]

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/benchmarks/groupby/group_scan_benchmark.cu [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  requests[0].aggregations.push_back(cudf::make_sum_aggregation<cudf::groupby_scan_aggregation>());
+
+  for (auto _ : state) {
+    cuda_event_timer timer(state, true);
+
+    auto result = gb_obj.scan(requests);
+  }
+}
+
+BENCHMARK_DEFINE_F(Groupby, BasicSumScan)(::benchmark::State& state) { BM_basic_sum_scan(state); }
```
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/benchmarks/groupby/group_scan_benchmark.cu:58]: `nvbench` requires target kernels to take an explicit stream (see [here](https://github.com/NVIDIA/nvbench/issues/13)) but CUDA stream is not exposed in the public `groupby::scan`. What I did in `JOIN_NVBENCH` is to expose stream arguments to hash join APIs. Do we want to do the same for `groupby::scan`? @jrhemstad Any comments?

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/benchmarks/groupby/group_scan_benchmark.cu:58]: I'd rather do this: https://github.com/NVIDIA/nvbench/issues/13

[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/benchmarks/groupby/group_scan_benchmark.cu:58]: OK. I will leave the new scan benchmark as gbench for this PR and look into https://github.com/NVIDIA/nvbench/issues/13.

**Review by karthikeyann — APPROVED** [review]

**Comment by PointKernel** [comment]
@gpucibot merge
